File: backends/python-service/app/services/fitness.py
from datetime import datetime
from app.models.types import FitnessData, FitnessSyncRequest
from app.services.cache import CacheService

class FitnessService:
    def __init__(self, cache: CacheService):
        self.cache = cache

    async def sync_fitness_data(self, req: FitnessSyncRequest) -> bool:
        """Sync fitness data from external sources"""
        fitness_key = f"fitness:{req.user_id}:{req.competition_id}:{req.date.strftime('%Y-%m-%d')}"
        
        fitness_data = FitnessData(
            id=f"{req.user_id}-{req.competition_id}-{int(req.date.timestamp())}",
            user_id=req.user_id,
            competition_id=req.competition_id,
            steps=req.steps,
            distance=req.distance,
            calories=req.calories,
            active_minutes=req.active_minutes,
            source=req.source,
            date=req.date,
            synced_at=datetime.utcnow(),
            created_at=datetime.utcnow()
        )
        
        await self.cache.set(fitness_key, fitness_data.dict(), 2592000)  # 30 days
        
        # Update aggregated stats
        await self._update_aggregated_stats(
            req.user_id,
            req.competition_id,
            fitness_data
        )
        
        return True

    async def get_user_stats(
        self, user_id: str, competition_id: str
    ) -> FitnessData:
        """Get aggregated fitness statistics for a user"""
        stats_key = f"fitness_stats:{user_id}:{competition_id}"
        
        stats = await self.cache.get(stats_key)
        
        if not stats:
            return FitnessData(
                id="",
                user_id=user_id,
                competition_id=competition_id,
                steps=0,
                distance=0.0,
                calories=0.0,
                active_minutes=0,
                source="",
                date=datetime.utcnow(),
                synced_at=datetime.utcnow(),
                created_at=datetime.utcnow()
            )
        
        return FitnessData(**stats)

    async def _update_aggregated_stats(
        self, user_id: str, competition_id: str, new_data: FitnessData
    ):
        """Update aggregated statistics for a user in a competition"""
        stats_key = f"fitness_stats:{user_id}:{competition_id}"
        
        current_stats = await self.cache.get(stats_key)
        
        if not current_stats:
            current_stats = {
                "id": "",
                "user_id": user_id,
                "competition_id": competition_id,
                "steps": 0,
                "distance": 0.0,
                "calories": 0.0,
                "active_minutes": 0,
                "source": new_data.source,
                "date": datetime.utcnow().isoformat(),
                "synced_at": datetime.utcnow().isoformat(),
                "created_at": datetime.utcnow().isoformat()
            }
        
        # Update aggregated values
        current_stats["steps"] += new_data.steps
        current_stats["distance"] += new_data.distance
        current_stats["calories"] += new_data.calories
        current_stats["active_minutes"] += new_data.active_minutes
        current_stats["synced_at"] = datetime.utcnow().isoformat()
        current_stats["source"] = new_data.source
        
        await self.cache.set(stats_key, current_stats, 2592000)  # 30 days

File: backends/go-service/internal/services/fitness.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/yourusername/health-competition-go/internal/models"
)

type FitnessService struct {
	cache       *CacheService
	supabaseURL string
}

func NewFitnessService(cache *CacheService, supabaseURL string) *FitnessService {
	return &FitnessService{
		cache:       cache,
		supabaseURL: supabaseURL,
	}
}

// SyncFitnessData syncs fitness data from external sources
func (s *FitnessService) SyncFitnessData(ctx context.Context, req *models.FitnessSyncRequest) error {
	// Store fitness data in cache
	fitnessKey := s.getFitnessDataKey(req.UserID, req.CompetitionID, req.Date)
	
	fitnessData := &models.FitnessData{
		ID:            fmt.Sprintf("%s-%s-%d", req.UserID, req.CompetitionID, req.Date.Unix()),
		UserID:        req.UserID,
		CompetitionID: req.CompetitionID,
		Steps:         req.Steps,
		Distance:      req.Distance,
		Calories:      req.Calories,
		ActiveMinutes: req.ActiveMinutes,
		Source:        req.Source,
		Date:          req.Date,
		SyncedAt:      time.Now(),
		CreatedAt:     time.Now(),
	}

	err := s.cache.Set(ctx, fitnessKey, fitnessData, 30*24*time.Hour)
	if err != nil {
		return err
	}

	// Update aggregated stats
	err = s.updateAggregatedStats(ctx, req.UserID, req.CompetitionID, fitnessData)
	if err != nil {
		return err
	}

	return nil
}

// GetUserStats retrieves aggregated fitness statistics for a user
func (s *FitnessService) GetUserStats(ctx context.Context, userID, competitionID string) (*models.FitnessData, error) {
	statsKey := s.getUserStatsKey(userID, competitionID)
	
	var stats models.FitnessData
	err := s.cache.Get(ctx, statsKey, &stats)
	if err != nil {
		// Return zero stats if not found
		return &models.FitnessData{
			UserID:        userID,
			CompetitionID: competitionID,
			Steps:         0,
			Distance:      0,
			Calories:      0,
			ActiveMinutes: 0,
		}, nil
	}

	return &stats, nil
}

// updateAggregatedStats updates the aggregated statistics for a user in a competition
func (s *FitnessService) updateAggregatedStats(ctx context.Context, userID, competitionID string, newData *models.FitnessData) error {
	statsKey := s.getUserStatsKey(userID, competitionID)

	// Get current stats
	var currentStats models.FitnessData
	err := s.cache.Get(ctx, statsKey, &currentStats)
	if err != nil {
		// Initialize if not exists
		currentStats = models.FitnessData{
			UserID:        userID,
			CompetitionID: competitionID,
			Steps:         0,
			Distance:      0,
			Calories:      0,
			ActiveMinutes: 0,
		}
	}

	// Update aggregated values
	currentStats.Steps += newData.Steps
	currentStats.Distance += newData.Distance
	currentStats.Calories += newData.Calories
	currentStats.ActiveMinutes += newData.ActiveMinutes
	currentStats.Source = newData.Source

	// Save updated stats
	err = s.cache.Set(ctx, statsKey, currentStats, 30*24*time.Hour)
	if err != nil {
		return err
	}

	return nil
}

// GetDailyStats retrieves fitness data for a specific day
func (s *FitnessService) GetDailyStats(ctx context.Context, userID, competitionID string, date time.Time) (*models.FitnessData, error) {
	fitnessKey := s.getFitnessDataKey(userID, competitionID, date)
	
	var fitnessData models.FitnessData
	err := s.cache.Get(ctx, fitnessKey, &fitnessData)
	if err != nil {
		return nil, err
	}

	return &fitnessData, nil
}

// Helper methods
func (s *FitnessService) getFitnessDataKey(userID, competitionID string, date time.Time) string {
	dateStr := date.Format("2006-01-02")
	return fmt.Sprintf("fitness:%s:%s:%s", userID, competitionID, dateStr)
}

func (s *FitnessService) getUserStatsKey(userID, competitionID string) string {
	return fmt.Sprintf("fitness_stats:%s:%s", userID, competitionID)
}
